sklearn/utils/tests/test_validation.py

```diff
@@ -937,6 +939,86 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
     assert type(raised_error.value) == type(err_msg)
 
 
+_psd_cases_valid = {
+    'nominal': ((1, 2), np.array([1, 2]), None, ""),
+    'nominal_np_array': (np.array([1, 2]), np.array([1, 2]), None, ""),
+    'insignificant_imag': ((5, 5e-5j), np.array([5, 0]), None, ""),
+    'significant neg': ((5, -1), np.array([5, 0]), PositiveSpectrumWarning,
+                        "There are significant negative eigenvalues"),
+    'significant neg float32': (np.array([3e-4, -2e-6], dtype=np.float32),
+                                np.array([3e-4, 0], dtype=np.float32),
+                                PositiveSpectrumWarning,
+                                "There are significant negative eigenvalues"),
+    'significant neg float64': (np.array([1e-5, -2e-10], dtype=np.float64),
+                                np.array([1e-5, 0], dtype=np.float64),
+                                PositiveSpectrumWarning,
+                                "There are significant negative eigenvalues"),
+    'insignificant neg': ((5, -5e-5), np.array([5, 0]), None, ""),
+    'insignificant neg float32': (np.array([1, -1e-6], dtype=np.float32),
+                                  np.array([1, 0], dtype=np.float32),
+                                  None, ""),
+    'insignificant neg float64': (np.array([1, -1e-10], dtype=np.float64),
+                                  np.array([1, 0], dtype=np.float64),
+                                  None, ""),
+}
+
+
+@pytest.mark.parametrize("lambdas, expected_lambdas, w_type, w_msg",
+                         list(_psd_cases_valid.values()),
+                         ids=list(_psd_cases_valid.keys()))
+@pytest.mark.parametrize("warn_on_zeros", [True, False])
+def test_check_psd_eigenvalues_valid(lambdas, expected_lambdas, w_type, w_msg,
+                                     warn_on_zeros):
+    # Test that ``_check_psd_eigenvalues`` returns the right output for valid
+    # input, possibly raising the right warning
+
+    with pytest.warns(w_type, match=w_msg) as w:
+        assert_array_equal(
+            _check_psd_eigenvalues(lambdas, warn_on_zeros=warn_on_zeros),
+            expected_lambdas
+        )
+    if w_type is None:
+        assert not w
+
+
+def test_check_psd_eigenvalues_bad_conditioning_warning():
+    # Test that ``_check_psd_eigenvalues`` raises a warning for bad
+    # conditioning when warn_on_zeros is set to True, and does not when it is
+    # set to False
+
+    input = (5, 4e-12)
+    output = np.array([5, 0])
+
+    with pytest.warns(None) as w:
+        assert_array_equal(_check_psd_eigenvalues(input, warn_on_zeros=False),
+                           output)
+    assert not w
+
+    w_msg = "the largest eigenvalue is more than 1.00E\\+12 times the smallest"
+    with pytest.warns(PositiveSpectrumWarning, match=w_msg) as w:
```

Comment:
not related to this PR. I think these warnings are exactly the ones which should be on `logging.warn` instead.

Reply:
(thanks for the review @adrinjalali  by the way !)

Very good point. see my comment below on the purpose of this. In the subsequent PR requiring this ( #12069 ) I anyway need to disable that warning because it is expected when the approximate decomposition methods for kPCA are used. So the choice is open to you as per this should be using `verbose` + `logging` or `warning`

Reply:
EDIT : I got confused between this PR and the subsequent PR. 

So in this PR we need to disable the warning. It is just left as "a possibility for the future".